lightning/src/ln/channelmanager.rs

```diff
@@ -2289,10 +2337,38 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 								let mut payment_secrets = self.pending_inbound_payments.lock().unwrap();
 								match payment_secrets.entry(payment_hash) {
 									hash_map::Entry::Vacant(_) => {
-										log_trace!(self.logger, "Failing new HTLC with payment_hash {} as we didn't have a corresponding inbound payment.", log_bytes!(payment_hash.0));
-										fail_htlc!(claimable_htlc);
+										match claimable_htlc.onion_payload {
+											OnionPayload::Invoice(_) => {
+												log_trace!(self.logger, "Failing new HTLC with payment_hash {} as we didn't have a corresponding inbound payment.", log_bytes!(payment_hash.0));
+												fail_htlc!(claimable_htlc);
+											},
+											OnionPayload::Spontaneous(preimage) => {
+												match channel_state.claimable_htlcs.entry(payment_hash) {
+													hash_map::Entry::Vacant(e) => {
+														e.insert(vec![claimable_htlc]);
+														new_events.push(events::Event::PaymentReceived {
+															payment_hash,
+															amt: amt_to_forward,
+															purpose: events::PaymentPurpose::SpontaneousPayment(preimage),
+														});
+													},
+													hash_map::Entry::Occupied(_) => {
+														log_trace!(self.logger, "Failing new keysend HTLC with payment_hash {} for a duplicative payment hash", log_bytes!(payment_hash.0));
```

Comment:
IIUC this case handling, if we've a valid (payment_secret, payment_hash) pair in `self.pending_inbound_payments` and we receive a latter keysend with the same hash, we fail the keysend instead of claiming ?

If you're an intermediary hop Mallory and you relay HTLC 1 hash X to final hop Bob, I think this let Mallory send a "final-receiver probing" keysend hash X to Bob and observe the early failure (`PendingHTLCsForwardable` vs `PaymentReceived` processing times diff) ? If this is correct, I'm not sure if it binds well the requirement laid out above ("Further, we must not expose whether we have any other HTLCs associated with the same payment_hash pending or not')...

Reply:
Grr good catch. Going one step up the stack, I wonder if we're gonna have users screw this up reliably too. If you're a user and doing accounting based on the `payment_hash` when you issue invoices, if you receive a keysend I can see users looking up the payment hash in their accounting database. I wonder if we should avoid exposing the keysend payment hash/preimage entirely. I guess some users may want them for other reasons, but we should go to great pains to point out in documentation that you MUST NEVER EVER co-mingle payment hash lookups between keysend and non-keysend. Maybe at least we should name the field differently to call it "keysend preimage" instead of "payment preimage"?

Reply:
Can we avoid to return payment hash to avoid mangle with the accounting database ? I can foresee users willingly to know the preimage if it has any semantic sense like a pay-to-publish, privacy-preserving satoshi place. But within the keysend model I would say the payment hash should be ignored by the receiver there is no hash validation operation to realize against a database.

And if the receiver wants to prove delivery they can still manually hash the preimage. We can even return a prefixed preimage by concatening a "Lightning keysend" ?

You might even have worst collision, where a low-value keysend trigger the release of a virtual/physical good by your business logic, substituting to a regular high-value payment...?

Reply:
Right, disregard, I had made that comment before I saw your suggestion at https://github.com/rust-bitcoin/rust-lightning/pull/967#discussion_r670539027